robot.py

```diff
@@ -0,0 +1,107 @@
+import asyncio
+from contextlib import asynccontextmanager
+
+from viam.components.board import Board
+from viam.components.servo import Servo
+from viam.robot.client import RobotClient
+from viam.rpc.dial import DialOptions
+
+
+@asynccontextmanager
+async def create_robot(creds, address):
+    """
+    This makes a connection to the hardware, creates a Robot object, and then
+    closes the connection when the context manager exits.
+    """
+    opts = RobotClient.Options(
+        refresh_interval=0,
+        dial_options=DialOptions(credentials=creds)
+    )
+    client = await RobotClient.at_address(address, opts)
+    servo = Servo.from_robot(client, "servo")
+    board = Board.from_robot(client, "pi")
+
+    robot = Robot(servo)
+    await robot.start()
+    try:
+        # TODO: don't return the board when we no longer need to.
+        yield robot, board
+    finally:
+        await robot.stop()
+        await client.close()
```

Comment:
I'm not entirely sure what this try/finally block does or how `yield` works with a context manager. Can you explain this to me?
I thought `yield` allowed a function to return something but keeps the state of the function, aka, the function has to be called again. So when it is called again, that's when the `finally` portion happens? Is it "called again" when the context manager is closed in `main()`?

Reply:
This is a magical thing about using the decorators from the `contextlib` library. They do some sneaky trick, to turn a generator that yields a single item (i.e., this function) into a context manager. Note the `@asynccontextmanager` decorator at the top of the function: it basically means that everything before the `yield` runs at the start of a `with` block, the yielded thing is the variable defined in the `with` statement itself, and everything after the `yield` runs when the `with` block ends. The `try...finally` is there because we want to clean up at the end even if something within the `with` block raised an exception (otherwise the exception would cause us to skip the rest of this function).

How does it work? :shrug: No idea. but once you recognize this pattern, it's way easier than implementing `__aenter__` and `__aexit__` and trying to remember the arguments for the latter.

FMI, see https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager (I've linked to the non-async version. The first line of the async version's documentation is "Similar to `contextmanager()`, but creates an asynchronous context manager.")

Reply:
Although I prefer using the `contextlib` decorators, I acknowledge that they're a little confusing the first time you see them, and part of their advantage (not needing to create a class when all you wanted was a function) doesn't help here (since we already needed to create a class). I'm open to undoing this and going back to `Robot.__aenter__` and `Robot.__aexit__` if you prefer.

Reply:
No, this is smart and I like it!